src/liballoc/btree/map.rs

```diff
@@ -2549,4 +2569,19 @@ impl<K: Ord, V, I: Iterator<Item = (K, V)>> Iterator for MergeIter<K, V, I> {
             }
         }
     }
+
+    // Currently unused. However, BTreeMap::from_sorted_iter may some day be
+    // rewritten to pre-allocate the correct amount of nodes, in which case
+    // having this will be helpful.
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        let (left_lower, left_upper) = self.left.size_hint();
+        let (right_lower, right_upper) = self.right.size_hint();
+        let lower = left_lower + right_lower;
+        let upper = match (left_upper, right_upper) {
+            (Some(left), Some(right)) => left.checked_add(right),
+            (left, None) => left,
+            (None, right) => right
```

Comment:
`None` as the upper is infinity, so I think this is wrong.

Also, this lift can often be written nicely with ?-on-Option, something like
```rust
(||{ left_upper?.checked_add(right_upper?) })()
```

Reply:
Yes, you are right! I got it backwards.

I don't think it's worth it to create a closure _purely_ to use `?`... it doesn't make it any more readable. IMHO a closure that is immediately called is harder to read & maintain.

Reply:
Well, it'll soon be `do catch { left_upper?.checked_add(right_upper?)? }` to avoid the IIFE (that I agree is ugly), but please don't do that until https://github.com/rust-lang/rust/pull/49371 merges.